### PR TITLE
Pass RDF `@id` to nodes as uuid.

### PR DESCRIPTION
**Pull request by marcenacp**
- Start deprecating `uid` to be replaced by `uuid`.
- Better hashing (as a consequence I discovered some issues in some datasets).

### CODE REVIEW

File: python/mlcroissant/mlcroissant/_src/core/uuid.py [hunk 1]
```diff
@@ -0,0 +1,18 @@
+"""Module to manipulate UUID."""
+
+import uuid
+
+from mlcroissant._src.core.types import Json
+
+
+def generate_uid() -> str:
+    """Generates a random UUID of version 4."""
```
Comment (ccl-core) [inline python/mlcroissant/mlcroissant/_src/core/uuid.py:9]: Maybe you could comment on why we choose version 4?

File: python/mlcroissant/mlcroissant/_src/datasets.py [hunk 2]
```diff
@@ -114,15 +117,16 @@ def __iter__(self):
...
         if self.debug:
             graphs_utils.pretty_print_graph(operations)
         # Downloads can be parallelized, so we execute them in priority.
         execute_downloads(operations)
         # We can stream the dataset iff the operation graph is a path graph (meaning
         # that all operations lie on a single straight line, i.e. have an
         # in-degree of 0 or 1. That means that the operation graph is a single line
-        # (without external joins for example).
+        # (without external joins for example). We only consider the operations that
+        # are useful to produce the `ReadFields(record_set)`.
```
Comment (ccl-core) [inline python/mlcroissant/mlcroissant/_src/datasets.py:129]: > We only consider the operations that are useful to produce the `ReadFields(record_set)`

This depends on _filter_interesting_operations() no?
If yes, then maybe this comment could be put above?

**Review by ccl-core — APPROVED** [review]
Thanks! I left a few NITs (mainly to help future readers to understand the technicalities)

File: python/mlcroissant/mlcroissant/_src/core/uuid.py (hunk 1 above)
Comment (marcenacp) [reply python/mlcroissant/mlcroissant/_src/core/uuid.py]: Done.

**Comment by marcenacp** [comment]
@mkuchnik The notebook `recipes/flores200_datapipes.ipynb` fails. It seems a part of the code still uses an older version of mlcroissant, because it complains that the field I just introduced (`uuid`) doesn't exist. Also, the notebook runs in local. Do you see why it could fail? Thanks.

**Comment by marcenacp** [comment]
I ended up implementing a simple `__deepcopy__` for all nodes and I created the issue https://github.com/mlcommons/croissant/issues/531 to track if there are any side effects.

**Pull request merged**